### DISCUSSION TITLE
Cannot save new login if personal ownership is disabled iOS Native app

### DISCUSSION THREAD

**Discussion by Gerardv514** [dani-garcia/vaultwarden#4767] [General]
@BlackDex @stefan0xC question, Bitwarden had a bug for a long time where a site couldn't be saved using an iOS device, share sheet from browser when on a website. At the time we just saw a spinning wheel and it wouldn't do anything. As per issue https://github.com/bitwarden/mobile/issues/1874 Bitwarden is stating it has been resolved in the new native apps. As per pull request https://github.com/bitwarden/mobile/pull/3265 it is still open (unless this is for the old apps - I'm a bit confused on what they did).

Is this something I would have to wait for Vaultwarden to bring in or is this strictly a client issue and should be working as they stated they have resolved in native apps? I tested this in native app and although it doesn't spin anymore, I now see a message stating an organizational policy is preventing you from saving items to your personal vault (on ios when adding an item thru browser using the share sheet and selecting bitwarden app).

**Comment by stefan0xC** [comment]
I'm not sure what they did either. Have you tested the new native app (which is currently [in an open beta](https://community.bitwarden.com/t/about-the-beta-program/39185)) if the issue still occurs?

**Reply by Gerardv514** [reply]
I did, I mentioned it in the post above of what I’m seeing now: “ I tested this in native app and although it doesn't spin anymore, I now see a message stating an organizational policy is preventing you from saving items to your personal vault (on ios when adding an item thru browser using the share sheet and selecting bitwarden app).”

The spin is gone but I cannot save because it thinks I’m saving to personal vault even though we have personal ownership disabled. 

So just trying to figure out if it’s something still on their end (that isn’t fixed or is still being worked on) or would changes have to be applied too on VW to get this to fully work.

**Reply by stefan0xC** [reply]
Which vaultwarden version are you using? `testing` or latest release?

**Reply by Gerardv514** [reply]
Latest

**Reply by stefan0xC** [reply]
Can you try with `testing` (just to make sure it's still a thing and not something that was caused by us, e.g. #4726).

Also from your description, I'm not sure what the issue exactly is. Can you maybe add screenshots and possibly logs? (If there are no server-side logs I'm afraid it will be a bit hard to find the issue or rather I'd suspect that the issue is caused client-side.)

**Comment by Gerardv514** [comment]
The issue exists even with testing.

Here is the workflow:

1. Organization with "single Organization" and "remove individual vault" on. (I believe this issue more pertains to the remove individual vault policy though.) 
2. Create a user (with user role) of an organization. I have assigned two collections to this user.
3. Using iOS iPhone with the native app installed.
4. Navigate to any website that has login fields, using a mobile browser such as Safari on the iPhone.
5. When on the login screen click the share sheet (square icon with the up arrow).
6. Find the bitwarden app in the list, the native app is listed as "Autofill with Bitwarden"
7. This will bring up login items, since this is a new site and we want to add to Bitwarden click the add + to add as a new item.
8. You'll see add item screen, with banner at top "An organization policy is affecting your ownership options" this is due to policy (no issue with this)
9. Scroll to the bottom, you'll see "Who owns this item?" which should be the organization.
10. Choose a collection to save to
11. Click save
    11a. Previously on the standard Bitwarden app when clicking save you would just get a spinning circle and nothing happens; this is 
             what they were fixing
    11b. Bitwarden mentioned in the issue that this issue was resolved in the native app.
12. In the native app when we hit save we receive an error has occurred. "Due to an Enterprise Policy, you are restricted from saving items to your personal vault."

We shouldn't get this error because we have remove individual (personal) vault ownership policy on. Furthermore, we can see the organization is the one that is owning this item.

I just also tested, opening up the native app and trying to add an item from there and I get the same error. So although bitwarden fixed the constant spinning circle, this second issue maybe a VW thing, maybe the code just hasn't come downstream yet? idk


![IMG_7089](https://github.com/user-attachments/assets/4175d92a-d404-44d7-891f-c5a164896f4e)
![IMG_7091](https://github.com/user-attachments/assets/2af0a494-a963-443f-9331-cbcf3b6e8167)

**Reply by stefan0xC** [reply]
Ah, okay. It could be related to the new flexible collections (or something else we have not implemented) but it's a bit hard to tell. I mean best thing to rule out a client-side bug would be if someone with a Bitwarden Enterprise license could reproduce your steps and confirm it actually works as expected on vault.bitwarden.com.

**Reply by Gerardv514** [reply]
I tried in the production app and the spinning is no longer present but i now get the same error. 

Only difference is current app doesn’t show the collections, whereas native apps do. 

Should we convert this to an issue here, and see what happens?

Unfortunately I don’t have Bw enterprise.

**Reply by Gerardv514** [reply]
Initially, with the spinning wheel and not able to save, this was deff something that they were reproduce. 

Now that they said it was/has been resolved in native and now seeing this message which we didn’t before I just don’t know if it’s still something on their end or VW. I would imagine though if it was them they would still see it because it’s reproducible when having remove personal vault policy on. 

Could be why the pull https://github.com/bitwarden/mobile/pull/3265  is still open. But again I’m confused as to why they closed the issue saying its resolved in native app.

**Reply by Gerardv514** [reply]
@stefan0xC looking at bitwarden’s response in the issue that was originally created for this, which includes a video recording it appears it’s working for them. 

Therefore this is probably an issue with Vaultwarden. Below is a link to the comment and video over there. 

https://github.com/bitwarden/mobile/issues/1874#issuecomment-2242299830

**Reply by stefan0xC** [reply]
@Gerardv514 I took a look at our code. Given that it seems to be returning the error message from https://github.com/dani-garcia/vaultwarden/blob/b27e6e30c9687d41e5596be73346ce0ed626068d/src/api/core/ciphers.rs#L334-L343 the issue might be caused here:  https://github.com/dani-garcia/vaultwarden/blob/b27e6e30c9687d41e5596be73346ce0ed626068d/src/api/core/ciphers.rs#L293 if the native client is sending everything in PascalCase instead of camelCase as mentioned in #4670 I think that at least `CipherData` would need to be fixed as well.

**Reply by BlackDex** [reply]
I do not know what those clients send, but it might be something different when we now except it to be.
As is with some other items also.

**Reply by stefan0xC** [reply]
Well, looking at the source code the requests from the iOS app (cf. [CipherRequestModel.swift](https://github.com/bitwarden/ios/blob/3ec242eb73735608fbed32837303b22ceb531abb/BitwardenShared/Core/Vault/Models/Request/CipherRequestModel.swift) and [CipherCreateRequestModel.swift](https://github.com/bitwarden/ios/blob/3ec242eb73735608fbed32837303b22ceb531abb/BitwardenShared/Core/Vault/Models/Request/CipherCreateRequestModel.swift)) appear to be in camelCase but I have no idea if that is what it actually sends. (The Android app appears to be more verbose in that regard, cf. [CipherJsonRequest](https://github.com/bitwarden/android/blob/62154f5261c14fc25731b7cbdaee0f8016d63e2a/app/src/main/java/com/x8bit/bitwarden/data/vault/datasource/network/model/CipherJsonRequest.kt) vs. [CreateCipherInOrganizationJsonRequest.kt](https://github.com/bitwarden/android/blob/62154f5261c14fc25731b7cbdaee0f8016d63e2a/app/src/main/java/com/x8bit/bitwarden/data/vault/datasource/network/model/CreateCipherInOrganizationJsonRequest.kt))

**Reply by Gerardv514** [reply]
I’m messing a bit more, I actually can’t do anything with any entries in the vault. Editing an existing item or even deleting an item doesn’t work and shows the same message. 

I can create new folders, but when attempting to move an existing vault item into the folder I again am prompted with the same message.

**Reply by Gerardv514** [reply]
I tested in web vault and no issue there, I’ll have to try the browser extension. 

I then turned off the policy “remove individual vault” and I am able to save and edit to the organization collection on the iOS native app. 

Not sure if that helps a bit more

**Reply by Gerardv514** [reply]
Actually now I’m getting this:

Tried a log out and log back in no dice. 

![IMG_7143](https://github.com/user-attachments/assets/9242a6a4-0797-4780-b9ef-1f8164193281)

**Reply by stefan0xC** [reply]
Ooh. That's bad. Sounds [like this issue](https://vaultwarden.discourse.group/t/error-must-provide-key/3835?u=stefan0xc) another user has reported. @BlackDex maybe we should warn users about using the beta apps (at least with iOS)?

**Reply by Gerardv514** [reply]
Agreed. Vault items no longer appear even in web vault, I just have a spinning circle. This is just a test container so I’m not worried about it. If I can help dig out anymore information for you guys I’d be happy to, just keep in mind all the backend stuff is a bit over my head. I’m also away until Thursday ish, so only have access thru my phone.

**Reply by BlackDex** [reply]
And this does not happen on the Bitwarden Cloud instance? 
Or you do not know this? 

Also, not sure how to warn all the people using the beta. It's called a beta for a reason of course. But this is nasty. 

We should check if this is a Vaultwarden specific item or only client.

**Reply by Gerardv514** [reply]
I do not know this for sure since I don’t have access to enterprise cloud, but as per bitwarden they are able to save and edit vault items with that policy on and using the native app. This is going by the video screenshot they posted in the git issue that was originally created (referenced earlier in this thread).

**Reply by Gerardv514** [reply]
I was able to restore using a snapshot. 

Aside from having the original issue, which is with my organizational policy on where users individual vault is removed, I am going to try and break this again and see if I can reproduce. 

It happened when trying to add an item or edit an item and/or  when/after I took off that organization policy. From the iOS app I am not able to save or edit anything.

**Comment by Gerardv514** [comment]
Ok here’s how to reproduce. It seems like the issue just happened when I created a new vault entry on iOS native app using the admin user and saving it to the organization, which worked. Keep in mind my regular user I was not able to save any items with because of the error stating due to org policy I cannot save to personal vault. After the admin saves that vault item I received that cryptography error in that admin vault. 

When this happened the first time I used the user account. I had turned off the policy “remove individual vault” then went back to the iOS native to the user account, and created a new entry. I recall coming to here to state that turning off that policy worked, meaning now I was able to save the entry using the iOS native app, but retracted because I received that cryptographic error.

**Comment by Gerardv514** [comment]
And just confirmed again, I went to the owner account of the org, turned off the remove individual vault policy. Then switch to iOS native app to user account with user role and created a new item in iOS native app. I did not receive an error on save but immediately after I received the cryptography error. Both of the broken vaults now no longer have access to any items in the vault.

**Reply by Gerardv514** [reply]
The cryptography issue is only presenting its self when saving an item to the org and not personal vault on iOS native app. Additionally if I have an item already in an org and I try to update it and save the cryptography issue arises on iOS native app.

**Reply by stefan0xC** [reply]
Thanks. I think the problem is that the iOS app calls it [`organizationID`](https://github.com/bitwarden/ios/blob/3ec242eb73735608fbed32837303b22ceb531abb/BitwardenShared/Core/Vault/Models/Request/CipherRequestModel.swift#L47) which is probably why the enforcement check fails but everything else does not (because it's an optional argument).

https://github.com/dani-garcia/vaultwarden/blob/b27e6e30c9687d41e5596be73346ce0ed626068d/src/api/core/ciphers.rs#L286-L288 is probably missing the reverse sanity check (if `data.collection_ids` is not empty, then `data.cipher.organization_id` should not be empty either) to prevent such issues from occurring. 

To actually fix the issues we probably need to add `#[serde(alias = "organizationID")]`  here https://github.com/dani-garcia/vaultwarden/blob/b27e6e30c9687d41e5596be73346ce0ed626068d/src/api/core/ciphers.rs#L211 so Vaultwarden does not overlook the provided organization id.

But that's only my guess from looking at the code. Due to a lack of iOS devices I can't actually test if this would actually fix the issue. (Which is why I'm a bit hesitant to write a PR for this.)

**Reply by Gerardv514** [reply]
Isn’t there some way to run some sort of virtual instance of iOS for developers?

**Reply by BlackDex** [reply]
If you have the right licenses and hardware maybe. It isn't as easy as starting Android studio which uses qemu to emulate an Android device on almost any platform.

You need a Mac to start with. And probably also a developer account maybe, not sure on the last part.

**Reply by Gerardv514** [reply]
Should we convert this to an issue, and label as help wanted specifically someone with an iOS device that can work along with?

**Reply by jedhamzawi** [reply]
@stefan0xC I can confirm that your proposed fix seemingly solves this problem! I stood up two Vaultwarden testing containers, one based on the current `main` branch and the other with the added serde alias attribute. For both containers, I created a new user and organization and added a password to the organization from the native iOS app. The container on `main` gave me the error @Gerardv514 showed, but the other container synced with no issues! See the below recordings:

`main` container:

https://github.com/user-attachments/assets/652ee101-fd1e-4a6d-af6a-510dd90f4994

https://github.com/user-attachments/assets/03da7795-b316-4a84-899d-b73c58f9f242

`test` (fixed) container:

https://github.com/user-attachments/assets/cc50ce80-3fd3-4f7b-bfb4-47e92ff9c693

https://github.com/user-attachments/assets/2a2528e0-7153-4ba3-84a8-3a091e66ebc6

And here's the git diff of what I changed for the `test` container:

```
@@ -208,6 +208,7 @@ pub struct CipherData {
     // Folder id is not included in import
     folder_id: Option<String>,
     // TODO: Some of these might appear all the time, no need for Option
+    #[serde(alias = "organizationID")]
     pub organization_id: Option<String>,

     key: Option<String>,
```

**Reply by jedhamzawi** [reply]
I also have a few related questions:

1. If the native Bitwarden app provided a JSON `CipherData` payload with a `organizationID` field, how could Vaultwarden have successfully deserialized the payload when it sees "invalid" fields?
2. There are other fields named in camelCase rather than snake_case in the iOS app (e.g. [password_history -> passwordHistory](https://github.com/bitwarden/ios/blob/3ec242eb73735608fbed32837303b22ceb531abb/BitwardenShared/Core/Vault/Models/Request/CipherRequestModel.swift#L85), [last_known_revision_date -> lastKnownRevisionDate](https://github.com/bitwarden/ios/blob/3ec242eb73735608fbed32837303b22ceb531abb/BitwardenShared/Core/Vault/Models/Request/CipherRequestModel.swift#L79). Will those propose a problem?
3. As far as I understand, it's probably very difficult (if not impossible) to repair a vault corrupted this way due to Bitwarden/Vaultwarden's encryption scheme (namely that encryption happens client-side and Vaultwarden stores encrypted blobs in its DB). There isn't a simple way to find the offending cipher and add the missing organization ID, is there?

Edit:

Ohhhh wait, I think I figured out 1 and 2: Vaultwarden already renames everything to camelCase, but the Bitwarden iOS app doesn't properly following camelCase conventions for `organization_id` and calls it `organizationID` rather than `organizationId`. Proving to me again why snake_case is superior 😆

**Reply by BlackDex** [reply]
Also, a lot of fields are part of the data field, which we just store as a blob/string without touching it (mostly).

And, we can also have some aliases for fields.

**Reply by BlackDex** [reply]
So, to fix it we would need to report an issue at Bitwarden i think, (not sure if it falls there now also) and add an alias.

**Reply by stefan0xC** [reply]
Thanks for testing @jedhamzawi. I've added a PR #4800 that should prevent such corruptions from occurring.

Regarding your third question: I don't think there's a way for us to fix corrupted entries because the organization_id is missing from the entries (making it rather impossible to somehow guess the right organization). Adding the organization id manually into the ciphers table might be enough to fix the issue but like you guessed it's not obvious which id is the correct one so it's probably better to just remove the offending entry and add it again.

**Reply by 11jwolfe2** [reply]
So I encountered this issue today. Assuming I have not updated anything else in my vault  would it be as easy as adding in the org_id to the cipher last updated in my db?

**Reply by stefan0xC** [reply]
@11jwolfe2 I have not tested it but I think it should be sufficient if you also know the correct `organization_id`.

**Reply by ibaivalencia** [reply]
> Thanks for testing @jedhamzawi. I've added a PR #4800 that should prevent such corruptions from occurring.
> 
> Regarding your third question: I don't think there's a way for us to fix corrupted entries because the organization_id is missing from the entries (making it rather impossible to somehow guess the right organization). Adding the organization id manually into the ciphers table might be enough to fix the issue but like you guessed it's not obvious which id is the correct one so it's probably better to just remove the offending entry and add it again.

Hi!
Sorry, I'm new with vaultwarden and I'm suffering from this bug. Can you point how to remove that last item I added to my vaultwarden to recover the functionality of my vault which is... currently dead? :-(

Many thanks!

**Reply by ibaivalencia** [reply]
> > Thanks for testing @jedhamzawi. I've added a PR #4800 that should prevent such corruptions from occurring.
> > Regarding your third question: I don't think there's a way for us to fix corrupted entries because the organization_id is missing from the entries (making it rather impossible to somehow guess the right organization). Adding the organization id manually into the ciphers table might be enough to fix the issue but like you guessed it's not obvious which id is the correct one so it's probably better to just remove the offending entry and add it again.
> 
> Hi! Sorry, I'm new with vaultwarden and I'm suffering from this bug. Can you point how to remove that last item I added to my vaultwarden to recover the functionality of my vault which is... currently dead? :-(
> 
> Many thanks!

Already have the answer to my question.
Open the sqlite database with sudo to gain write rights:
 sudo sqlite3 db.sqlite3
Query the last created item:
 SELECT uuid, created_at, updated_at from ciphers order by created_at asc limit 1;
Remove the last item, getting the id:
 DELETE from ciphers where uuid='2c7a4bb5-0100-46ba-9a73-35c13aa4b347';

My vaultwarden is back again!
